graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/caching/CacheControlInstrumentation.java

```diff
@@ -0,0 +1,372 @@
+      if (policy.maxAge.isPresent()
+          && (!maxAge.isPresent() || policy.maxAge.get() < maxAge.get())) {
+        this.maxAge = policy.maxAge;
+      }
+
+      if (policy.scope.isPresent()
+          && (!scope.isPresent() || !scope.get().equals(CacheControlScope.PRIVATE))) {
+        this.scope = policy.scope;
+      }
+    }
+
+    void restrict(Hint hint) {
+      if (hint.maxAge.isPresent() && (!maxAge.isPresent() || hint.maxAge.get() < maxAge.get())) {
+        this.maxAge = hint.maxAge;
+      }
+
+      if (hint.scope.isPresent()
+          && (!scope.isPresent() || !scope.get().equals(CacheControlScope.PRIVATE))) {
+        this.scope = hint.scope;
+      }
+    }
+
+    void restrict(Integer maxAge) {
+      this.maxAge = Optional.of(maxAge);
+    }
+
+    void replace(Hint hint) {
+      if (hint.maxAge.isPresent()) {
+        this.maxAge = hint.maxAge;
+      }
+
+      if (hint.scope.isPresent()) {
+        this.scope = hint.scope;
+      }
+    }
+
+    void replace(Optional<CacheControlScope> scope) {
+      if (scope.isPresent()) {
+        this.scope = scope;
+      }
+    }
+
+    public Optional<String> getPolicy() {
+      Integer maxAgeValue = maxAge.orElse(0);
+      if (maxAgeValue.equals(0)) {
+        return Optional.empty();
+      }
+      return Optional.of(
+          String.format(
+              "max-age=%d, %s",
+              maxAgeValue, scope.orElse(CacheControlScope.PUBLIC).toString().toLowerCase()));
+    }
+
+    public boolean hasMaxAge() {
+      return maxAge.isPresent();
+    }
+
+    public boolean hasScope() {
+      return scope.isPresent();
+    }
+  }
+
+  private static class HintCache {
+    private final Map<GraphQLCompositeType, Hint> typeHintCache = new HashMap<>();
+    private final Map<GraphQLFieldDefinition, Hint> fieldHintCache = new HashMap<>();
+
+    public Optional<Hint> getHint(GraphQLFieldDefinition fieldDefinition) {
+      if (fieldHintCache.containsKey(fieldDefinition)) {
+        return Optional.of(fieldHintCache.get(fieldDefinition));
+      }
+
+      Optional<Hint> fieldHint = Hint.fromDirective(fieldDefinition.getDirective(DIRECTIVE_NAME));
+      if (fieldHint.isPresent()) {
+        fieldHintCache.put(fieldDefinition, fieldHint.get());
+        return fieldHint;
+      }
+
+      return Optional.empty();
+    }
+
+    public Optional<Hint> getHint(GraphQLCompositeType type) {
+      if (!(type instanceof GraphQLDirectiveContainer)) {
+        return Optional.empty();
+      }
+
+      GraphQLDirectiveContainer directiveContainer = (GraphQLDirectiveContainer) type;
+
+      if (typeHintCache.containsKey(type)) {
+        return Optional.of(typeHintCache.get(type));
+      }
+
+      Optional<Hint> hint = Hint.fromDirective(directiveContainer.getDirective(DIRECTIVE_NAME));
+      if (hint.isPresent()) {
+        typeHintCache.put(type, hint.get());
+        return hint;
+      }
+
+      return Optional.empty();
+    }
+  }
+
+  private static class Hint {
```

Comment:
I like `CacheControlDirective`, it's a "user-friendly" representation of a `GraphQLDirective`

I also like `CacheHint` because it maps the [documentation](https://www.apollographql.com/docs/apollo-server/performance/caching/#in-your-resolvers-dynamic)

Reply:
went with `CacheControlDirective`!